File: android/src/main/kotlin/io/glassfy/glassfy_flutter/GlassfyFlutterPlugin.kt
package io.glassfy.glassfy_flutter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.annotation.NonNull
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.glassfy.glue.GlassfyGlue
import io.glassfy.paywall.GlassfyPaywall
import org.json.JSONObject

/**
 * GlassfyFlutterPlugin
 */
class GlassfyFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var activity: Activity
    private var fragmentActivity: FragmentActivity? = null
    private var paywallListener: FlutterPaywallListener? = null
    private lateinit var paywallEventsChannel: EventChannel
    private var paywallEventsSink: EventChannel.EventSink? = null
    private var paywallFragment: DialogFragment? = null

    override fun onDetachedFromActivity() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        fragmentActivity= binding.activity as? FragmentActivity
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    private fun pluginCompletion(@NonNull result: Result, value: String?, error: String?) {
        if (error != null) {
            result.error(error, null, null)
            return
        }
        result.success(value)
    }

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        setupMethodChannel(flutterPluginBinding)
        setupPaywallEventsChannel(flutterPluginBinding)
    }

    private fun setupMethodChannel(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "glassfy_flutter")
        channel.setMethodCallHandler(this)
    }

    private fun setupPaywallEventsChannel(binding: FlutterPlugin.FlutterPluginBinding) {
        paywallEventsChannel = EventChannel(binding.binaryMessenger, "paywallEvent")
        paywallEventsChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, sink: EventChannel.EventSink) {
                paywallEventsSink = sink
            }

            override fun onCancel(arguments: Any?) {
                paywallEventsSink = null
            }
        })
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {

        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "sdkVersion" -> {
                GlassfyGlue.sdkVersion { v, e -> pluginCompletion(result, v, e) }
            }

            "initialize" -> {
                val framework: String = "flutter";
                val apiKey: String = call.argument("apiKey") ?: ""
                val watcherMode: Boolean = call.argument("watcherMode") ?: false
                val version: String = call.argument("version") ?: "unknown"
                GlassfyGlue.initialize(context, apiKey, watcherMode, framework, version) { v, e ->
                    pluginCompletion(result, v, e)
                }
            }

            "setLogLevel" -> {
                val logLevel: Int = call.argument("logLevel") ?: 0
                GlassfyGlue.setLogLevel(logLevel)
            }

            "offerings" -> {
                GlassfyGlue.offerings() { v, e -> pluginCompletion(result, v, e) }
            }

            "purchaseHistory" -> {
                GlassfyGlue.purchaseHistory() { v, e -> pluginCompletion(result, v, e) }
            }

            "permissions" -> {
                GlassfyGlue.permissions() { v, e -> pluginCompletion(result, v, e) }
            }

            "skuWithId" -> {
                val identifier: String = call.argument("identifier") ?: ""
                GlassfyGlue.skuWithId(identifier) { v, e -> pluginCompletion(result, v, e) }
            }

            "skuWithIdAndStore" -> {
                val identifier: String = call.argument("identifier") ?: ""
                val store: Int = call.argument("store") ?: 0
                GlassfyGlue.skuWithIdAndStore(identifier, store) { v, e ->
                    pluginCompletion(
                        result,
                        v,
                        e
                    )
                }
            }

            "purchaseSku" -> {
                val sku: HashMap<String, String>? = call.argument("sku")
                val skuToUpgrade: HashMap<String, Any>? = call.argument("skuToUpgrade")
                val skuId = sku?.get("skuId");
                var subscriptionUpdateId: String? = null
                var proration: Int? = null
                if (skuToUpgrade != null) {
                    subscriptionUpdateId = skuToUpgrade.get("skuId") as String?;
                    if (subscriptionUpdateId == null) {
                        result.error("Invalid skuToUpgrade", null, null)
                        return
                    }
                    proration = call.argument("prorationMode")
                }
                if (skuId == null) {
                    result.error("Invalid SKU", null, null)
                    return
                }

                GlassfyGlue.purchaseSku(
                    activity,
                    skuId,
                    subscriptionUpdateId,
                    proration
                ) { v, e -> pluginCompletion(result, v, e) }
            }

            "restorePurchases" -> {
                GlassfyGlue.restorePurchases() { v, e -> pluginCompletion(result, v, e) }
            }

            "setDeviceToken" -> {
                // not applicable in Android
            }

            "setEmailUserProperty" -> {
                val email: String = call.argument("email") ?: ""
                GlassfyGlue.setEmailUserProperty(email) { v, e -> pluginCompletion(result, v, e) }
            }

            "setExtraUserProperty" -> {
                val extra: Map<String, String>? = call.argument("extraProp")
                GlassfyGlue.setExtraUserProperty(extra) { v, e -> pluginCompletion(result, v, e) }
            }

            "getExtraUserProperty" -> {
                GlassfyGlue.getExtraUserProperty() { v, e -> pluginCompletion(result, v, e) }
            }

            "connectCustomSubscriber" -> {
                val subscriberId: String? = call.argument("subscriberId")
                GlassfyGlue.connectCustomSubscriber(subscriberId) { v, e ->
                    pluginCompletion(
                        result,
                        v,
                        e
                    )
                }
            }

            "connectPaddleLicenseKey" -> {
                val licenseKey: String = call.argument("licenseKey") ?: ""
                val force: Boolean = call.argument("force") ?: false
                GlassfyGlue.connectPaddleLicenseKey(licenseKey, force) { v, e ->
                    pluginCompletion(
                        result,
                        v,
                        e
                    )
                }
            }

            "connectGlassfyUniversalCode" -> {
                val universalCode: String = call.argument("universalCode") ?: ""
                val force: Boolean = call.argument("force") ?: false
                GlassfyGlue.connectGlassfyUniversalCode(
                    universalCode,
                    force
                ) { v, e -> pluginCompletion(result, v, e) }
            }

            "setAttribution" -> {
                val type: Int = call.argument("type") ?: -1
                val value: String = call.argument("value") ?: ""
                GlassfyGlue.setAttribution(type, value) { v, e ->
                    pluginCompletion(result, v, e)
                }
            }

            "setAttributions" -> {
                val items: List<Map<String, Any?>> = call.argument("items") ?: listOf()
                GlassfyGlue.setAttributions(items) { v, e -> pluginCompletion(result, v, e) }
            }

            "showPaywall" -> {
                val fragmentActivity = this.fragmentActivity
                val remoteConfig: String = call.argument("remoteConfig") ?: ""
                val awaitLoading: Boolean = call.argument("awaitLoading") ?: false

                if (paywallFragment != null) {
                    result.error(
                        "Only one paywall can be shown at a time, please call `GlassfyPaywall.close()`",
                        null,
                        null
                    )
                    return
                }
                if (fragmentActivity == null) {
                    result.error(
                        "Could not find a FragmentActivity instance please check our documentation",
                        null,
                        null
                    )
                    return
                }

                val listener = FlutterPaywallListener { eventName, payload ->
                    sendEvent(eventName, payload)
                }

                GlassfyPaywall.fragment(remoteConfig, awaitLoading) { paywall, error ->
                    if (paywall != null) {
                        paywall?.setCloseHandler(listener.onClose)
                        paywall?.setPurchaseHandler(listener.onPurchase)
                        paywall?.setRestoreHandler(listener.onRestore)
                        paywall?.setLinkHandler(listener.onLink)
                        paywall?.show(fragmentActivity.supportFragmentManager, "paywall")
                        paywallFragment = paywall
                        paywallListener = listener
                        result.success(null)
                    } else {
                        result.error("Failed to show paywall", error?.toString(), null)
                    }
                }
            }

            "closePaywall" -> {
                closePaywall()
            }

            "openUrl" -> {
                try {
                    val urlString: String = call.argument("url") ?: ""
                    val i = Intent(Intent.ACTION_VIEW).apply { data = Uri.parse(urlString) }
                    activity?.startActivity(i)
                    result.success("")
                } catch (e: Exception) {
                    result.error(e.toString(), null, null)
                }
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    internal fun closePaywall() {
        activity.runOnUiThread {
            paywallFragment?.dismiss()
            paywallFragment = null
            paywallListener = null
        }
    }

    private fun sendEvent(name: String, payload: JSONObject) {
        activity.runOnUiThread {
            val event = JSONObject()
            event.put("name", name)
            event.put("payload", payload)
            paywallEventsSink?.success(event.toString())
        }
    }
}

File: ios/Classes/GlassfyFlutterPlugin.m
#import "GlassfyFlutterPlugin.h"
#import "FlutterPaywallListener.h"

@interface GlassfyFlutterPlugin()
@property (nonatomic, strong) FlutterEventSink paywallEventSink;
@property (nonatomic, strong) FlutterEventChannel* paywallEventChannel;
@property (nonatomic, strong) GYPaywallViewController* _Nullable paywallViewController;
@property (nonatomic, strong) FlutterPaywallListener* _Nullable paywallListener;
@end

@implementation GlassfyFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel =
    [FlutterMethodChannel methodChannelWithName:@"glassfy_flutter"
                                binaryMessenger:[registrar messenger]];
    GlassfyFlutterPlugin *instance = [[GlassfyFlutterPlugin alloc] init];
    instance.channel = channel;
    [registrar addMethodCallDelegate:instance channel:channel];
    
    instance.paywallEventChannel = [FlutterEventChannel
                                    eventChannelWithName:@"paywallEvent"
                                    binaryMessenger:[registrar messenger]];
    [instance.paywallEventChannel setStreamHandler:instance];
}

- (FlutterError*)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events {
    self.paywallEventSink = events;
    return nil;
}

- (FlutterError*)onCancelWithArguments:(id)arguments {
    self.paywallEventSink = nil;
    return nil;
}

NSString *GlassfyDidPurchaseEventFromDelegate = @"gy_did_purchase_product";

- (void)handleMethodCall:(FlutterMethodCall *)call
                  result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    
    if ([@"getPlatformVersion" isEqualToString:call.method]) {
        result([@"VER "
                stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    } else if ([@"sdkVersion" isEqualToString:call.method]) {
        [GlassfyGlue
         sdkVersionWithCompletion:[self
                                   convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"initialize" isEqualToString:call.method]) {
        NSString* apiKey = arguments[@"apiKey"];
        NSString* version = arguments[@"version"];
        BOOL watcherMode = [arguments[@"watcherMode"] boolValue];
        [GlassfyGlue
         initializeWithApiKey: apiKey
         watcherMode: watcherMode
         crossPlatformSdkFramework: @"flutter"
         crossPlatformSdkVersion: version
         withCompletion: [self convertGlassfyGlueResultToFlutter:result]];
        [GlassfyGlue setPurchaseDelegate:self];
    } else if ([@"setLogLevel" isEqualToString:call.method]) {
        int logLevel = [arguments[@"logLevel"] intValue];
        [GlassfyGlue setLogLevel:logLevel];
    } else if ([@"offerings" isEqualToString:call.method]) {
        [GlassfyGlue offeringsWithCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"purchaseHistory" isEqualToString:call.method]) {
        [GlassfyGlue purchaseHistoryWithCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"skuWithId" isEqualToString:call.method]) {
        NSString *identifier = arguments[@"identifier"];
        [GlassfyGlue skuWithId:identifier
                withCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"skuWithIdAndStore" isEqualToString:call.method]) {
        NSString *identifier = arguments[@"identifier"];
        int store = [arguments[@"store"] intValue];
        [GlassfyGlue skuWithId:identifier
                         store:store
                    completion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"login" isEqualToString:call.method]) {
        NSString *userid = arguments[@"userid"];
        [GlassfyGlue
         connectCustomSubscriber:userid
         completion:[self
                     convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"logout" isEqualToString:call.method]) {
        [GlassfyGlue
         connectCustomSubscriber:nil
         completion:[self
                     convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"permissions" isEqualToString:call.method]) {
        [GlassfyGlue permissionsWithCompletion:
         [self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"purchaseSku" isEqualToString:call.method]) {
        NSDictionary *sku = arguments[@"sku"];
        [GlassfyGlue purchaseSku:sku
                  withCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"restorePurchases" isEqualToString:call.method]) {
        [GlassfyGlue restorePurchasesWithCompletion:
         [self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"setDeviceToken" isEqualToString:call.method]) {
        NSString *token = arguments[@"token"];
        [GlassfyGlue
         setDeviceToken:token
         withCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"setEmailUserProperty" isEqualToString:call.method]) {
        NSString *email = arguments[@"email"];
        [GlassfyGlue
         setEmailUserProperty:email
         withCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"setExtraUserProperty" isEqualToString:call.method]) {
        NSDictionary *extraProp = arguments[@"extraProp"];
        [GlassfyGlue
         setExtraUserProperty:extraProp
         withCompletion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"getExtraUserProperty" isEqualToString:call.method]) {
        [GlassfyGlue getExtraUserPropertyWithCompletion:
         [self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"connectCustomSubscriber" isEqualToString:call.method]) {
        NSString *subscriberId = arguments[@"subscriberId"];
        [GlassfyGlue
         connectCustomSubscriber:subscriberId
         completion:[self
                     convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"connectPaddleLicenseKey" isEqualToString:call.method]) {
        NSString *licenseKey = arguments[@"licenseKey"];
        BOOL force = [arguments[@"force"] boolValue];
        [GlassfyGlue
         connectPaddleLicenseKey:licenseKey
         force:force
         completion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"connectGlassfyUniversalCode" isEqualToString:call.method]) {
        NSString *universalCode = arguments[@"universalCode"];
        BOOL force = [arguments[@"force"] boolValue];
        [GlassfyGlue
         connectGlassfyUniversalCode:universalCode
         force:force
         completion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"setAttribution" isEqualToString:call.method]) {
        NSNumber *type = arguments[@"type"];
        NSString *value = arguments[@"value"];
        [GlassfyGlue setAttributionType:type
                                  value:value
                             completion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"setAttributions" isEqualToString:call.method]) {
        NSArray *items = arguments[@"items"];
        [GlassfyGlue setAttributions:items
                          completion:[self convertGlassfyGlueResultToFlutter:result]];
    } else if ([@"openUrl" isEqualToString:call.method]) {
        NSString *urlString = arguments[@"url"];
        NSURL *url = [NSURL URLWithString:urlString];
        if (url) {
            [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:nil];
        }
        result(@"");
    } else if ([@"showPaywall" isEqualToString:call.method]) {
        NSString *remoteConfig = arguments[@"remoteConfig"];
        BOOL awaitLoading = arguments[@"awaitLoading"];
        [self showPaywallWithRemoteConfig:remoteConfig awaitLoading:awaitLoading result:result];
    } else if ([@"closePaywall" isEqualToString:call.method]) {
        if (self.paywallViewController != nil) {
            [self.paywallViewController dismissViewControllerAnimated:YES completion:nil];
        }
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)showPaywallWithRemoteConfig:(NSString *)remoteConfig
                       awaitLoading:(BOOL)awaitLoading
                             result:(FlutterResult)result {
    [Glassfy paywallViewControllerWithRemoteConfigurationId:remoteConfig
                                               awaitLoading:awaitLoading
                                                 completion:^(GYPaywallViewController * _Nullable viewController, NSError * _Nullable error) {
        if (error != nil) {
            result(error);
            return;
        }
        self.paywallViewController = viewController;
        
        self.paywallListener = [[FlutterPaywallListener alloc] initWithSendEvent:^(NSString * _Nullable eventName, NSDictionary * _Nullable payload) {
            [self sendEventWithName:eventName payload:payload];
        }];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            UIViewController *rootViewController = UIApplication.sharedApplication.delegate.window.rootViewController;
            [rootViewController presentViewController:viewController animated:YES completion:nil];
        });
        
        [viewController setCloseHandler:^(GYTransaction *transaction, NSError *error) {
            [self.paywallListener onCloseWithTransaction:transaction error:error];
        }];
        [viewController setLinkHandler:^(NSURL *link) {
            [self.paywallListener onLinkWithURL:link];
        }];
        [viewController setRestoreHandler:^{
            [self.paywallListener onRestore];
        }];
        [viewController setPurchaseHandler:^(GYSku *sku) {
            [self.paywallListener onPurchaseWithSKU:sku];
        }];
        
        result(@{@"result": @"success"});
    }];
}

- (void (^)(NSDictionary *,
            NSError *))convertGlassfyGlueResultToFlutter:(FlutterResult)result {
    return ^(NSDictionary *_Nullable resultDictionary, NSError *_Nullable error) {
        if (error) {
            FlutterError *ferror = [FlutterError
                                    errorWithCode:[NSString stringWithFormat:@"%ld", (long)error.code]
                                    message:error.localizedDescription
                                    details:error.description];
            result(ferror);
        } else {
            if (resultDictionary == nil) {
                result(nil);
                return;
            }
            NSError *error;
            NSData *jsonData = [NSJSONSerialization dataWithJSONObject:resultDictionary options:0 error:&error];
            if (error) {
                result([FlutterError
                        errorWithCode:[NSString stringWithFormat:@"%ld", (long)error.code]
                        message:error.localizedDescription
                        details:error.description]);
                return;
            }
            
            NSString *jsonString =
            [[NSString alloc] initWithData:jsonData
                                  encoding:NSUTF8StringEncoding];
            result(jsonString);
        }
    };
}

- (void)sendEventWithName:(NSString*)name payload:(NSDictionary*)payload {
    if (self.paywallEventSink) {
        NSMutableDictionary* event = [NSMutableDictionary dictionary];
        event[@"name"] = name;
        event[@"payload"] = payload;
        
        NSError *error;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:event options:NSJSONWritingPrettyPrinted error:&error];
        
        if (jsonData) {
            NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
            self.paywallEventSink(jsonString);
        }
    }
}

#pragma mark -
#pragma mark Glassfy Delegate Methods

- (void)didPurchaseProduct:(NSDictionary<NSString *, id> *)transaction {
    
    NSError *error;
    NSData *jsonData = [NSJSONSerialization
                        dataWithJSONObject:transaction
                        options:0 // Pass 0 if you don't care about the readability
                        // of the generated string
                        error:&error];
    if (error) {
        return;
    }
    
    NSString *jsonString =
    [[NSString alloc] initWithData:jsonData
                          encoding:NSUTF8StringEncoding];
    [self.channel invokeMethod:GlassfyDidPurchaseEventFromDelegate
                     arguments:jsonString];
}

- (void)handlePromotedProductId:(NSString *)productid
              withPromotionalId:(NSString *)promoid
                purchaseHandler:(void (^)(GYPaymentTransactionBlock))purchase {
}

@end
